src/anyio/from_thread.py

```diff
@@ -51,12 +65,21 @@ def run_sync(func: Callable[..., T_Retval], *args: object) -> T_Retval:
     :return: the return value of the callable
 
     """
-    try:
-        asynclib = threadlocals.current_async_module
-    except AttributeError:
-        raise RuntimeError('This function can only be run from an AnyIO worker thread')
+    if asynclib_token is None:
+        try:
+            asynclib = threadlocals.current_async_module
+        except AttributeError:
+            raise RuntimeError('This function can only be run from an AnyIO worker thread')
 
-    return asynclib.run_sync_from_thread(func, *args)
+        return asynclib.run_sync_from_thread(func, *args)
+
+    if isinstance(asynclib_token, AbstractEventLoop):
+        return get_asynclib("asyncio").run_sync_from_thread(func, *args, loop=asynclib_token)
+
+    if type(asynclib_token).__module__.startswith("trio."):
+        return get_asynclib("trio").run_sync_from_thread(func, *args, trio_token=asynclib_token)
```

Comment:
I think that the fundamental problem here is that tokens aren't being treated as opaque AnyIO specific objects here. If we did that, we could embed the backend name in them and not have to write code like this.

Reply:
I think it's worth it to be able to accept a trio.lowlevel.current_token() here

Reply:
I don't agree with you here. Just as trio has its own opaque tokens, AnyIO could have its own. The fact that the "token" involves an asyncio event loop object or a trio token is just an implementation detail.